File: rfcs/bench-0104/harness/js/revcomp.js
// reverse-complement, the JavaScript leg of RFC-0104 M2.
//
// The specification is examples/revcomp.vyrn. Same 256-entry complement table,
// same 60-column wrap, byte-identical output.
//
//   $ node revcomp.js < ../../fasta-1000.expected
//
// Structural differences from the Vyrn program:
//
//   * stdin is read once with fs.readFileSync(0) and split on '\n'. The Vyrn
//     program pulls a line at a time from `readLine()`, so it never holds the
//     whole input; this leg does. The transformation itself is identical, and
//     both hold one whole sequence at a time regardless.
//   * a sequence is a JavaScript string, not an Array<UInt8>, so there is no
//     `bytes(l)` allocation per input line and no `stringFromBytes` per output
//     line. That UTF-8 validation is work this leg does not do.

'use strict';
const fs = require('fs');

let pending = '';
function emit(s) {
    pending += s;
    if (pending.length >= 1 << 20) flush();
}
function flush() {
    if (pending.length === 0) return;
    const buf = Buffer.from(pending, 'latin1');
    pending = '';
    let off = 0;
    while (off < buf.length) {
        try {
            off += fs.writeSync(1, buf, off, buf.length - off);
        } catch (e) {
            if (e.code !== 'EAGAIN') throw e;
        }
    }
}

/// The IUB complement table, 256 entries, indexed by the input byte. Bases that
/// are not IUB codes map to themselves, which is what makes the lookup total.
function complementTable() {
    const t = [];
    let i = 0;
    while (i < 256) {
        t.push(String.fromCharCode(i));
        i = i + 1;
    }
    const from = 'ACBDGHKMNSRUTWVYacbdghkmnsrutwvy';
    const to = 'TGVHCDMKNSYAAWBRTGVHCDMKNSYAAWBR';
    let k = 0;
    while (k < from.length) {
        t[from.charCodeAt(k)] = to[k];
        k = k + 1;
    }
    return t;
}

/// `seq` backwards, every base complemented, printed 60 columns to a line with a
/// short last line if it does not divide.
function writeReverseComplement(seq, table) {
    let w = '';
    let i = seq.length - 1;
    while (i >= 0) {
        w += table[seq.charCodeAt(i)];
        if (w.length === 60) {
            emit(w + '\n');
            w = '';
        }
        i = i - 1;
    }
    if (w.length > 0) {
        emit(w + '\n');
    }
}

/// The whole run: one header and one sequence at a time, transformed and written
/// as soon as its last line arrives.
function run() {
    const table = complementTable();
    const lines = fs.readFileSync(0, 'latin1').split('\n');
    let header = '';
    let parts = [];
    for (const l of lines) {
        if (l.length > 0 && l[0] === '>') {
            if (header.length > 0) {
                emit(header + '\n');
                writeReverseComplement(parts.join(''), table);
                parts = [];
            }
            header = l;
        } else {
            parts.push(l);
        }
    }
    if (header.length > 0) {
        emit(header + '\n');
        writeReverseComplement(parts.join(''), table);
    }
}

run();
flush();

File: rfcs/bench-0104/harness/rust/revcomp.rs
// reverse-complement, from the Computer Language Benchmarks Game.
//
// The Rust leg of RFC-0104 M2. It computes what `examples/revcomp.vyrn`
// computes and prints the same bytes.
//
//   $ rustc -C opt-level=3 -o revcomp.exe revcomp.rs
//   $ ./revcomp < ../../fasta-1000.expected
//
// Safe Rust, std only. There is no N: the input is FASTA on stdin.

use std::io::{self, BufRead, BufWriter, Write};

/// The IUB complement table, 256 entries, indexed by the input byte. Bases that
/// are not IUB codes map to themselves, which is what makes the lookup total.
fn complement_table() -> [u8; 256] {
    let mut t = [0u8; 256];
    for i in 0..256 {
        t[i] = i as u8;
    }
    let from = b"ACBDGHKMNSRUTWVYacbdghkmnsrutwvy";
    let to = b"TGVHCDMKNSYAAWBRTGVHCDMKNSYAAWBR";
    for k in 0..from.len() {
        t[from[k] as usize] = to[k];
    }
    t
}

/// `seq` backwards, every base complemented.
fn reverse_complement(seq: &[u8], table: &[u8; 256]) -> Vec<u8> {
    seq.iter().rev().map(|&b| table[b as usize]).collect()
}

/// `bs` printed 60 columns to a line, with a short last line if it does not
/// divide.
fn write_wrapped(out: &mut impl Write, bs: &[u8]) {
    for chunk in bs.chunks(60) {
        out.write_all(chunk).unwrap();
        out.write_all(b"\n").unwrap();
    }
}

/// The whole run: one header and one sequence at a time, transformed and
/// written as soon as its last line arrives.
fn run(input: &mut impl BufRead, out: &mut impl Write) {
    let table = complement_table();
    let mut header = String::new();
    let mut seq: Vec<u8> = Vec::new();
    for line in input.lines() {
        let l = line.unwrap();
        if l.as_bytes().first() == Some(&b'>') {
            if !header.is_empty() {
                writeln!(out, "{}", header).unwrap();
                write_wrapped(out, &reverse_complement(&seq, &table));
                seq.clear();
            }
            header = l;
        } else {
            seq.extend_from_slice(l.as_bytes());
        }
    }
    if !header.is_empty() {
        writeln!(out, "{}", header).unwrap();
        write_wrapped(out, &reverse_complement(&seq, &table));
    }
}

fn main() {
    let stdin = io::stdin();
    let mut input = stdin.lock();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    run(&mut input, &mut out);
}
